text/0000-usize-panic-overflow.md

```diff
@@ -0,0 +1,185 @@
+- Feature Name: `usize_panic_on_overflow`
+- Start Date: 2019-02-02
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+In `release` builds, adopt the current `debug_assertions` behavior of panicking
+on overflow for arithmetic with `usize` (but not other sizes/types of integers).
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's motivations is to make programming safer, particularly against
+memory corruption vulnerabilities. In safe Rust code, integer overflows do not
+lead to memory unsafety, however when combined with `unsafe`, integer overflow
+is a frequent source of security vulnerabilities. We consider four examples:
+
+- [CVE-2018-1000810](https://groups.google.com/d/msg/rustlang-security-announcements/CmSuTm-SaU0/AzVznVcTCgAJ):
+  Integer overflow leading to heap buffer overflow in `str::repeat`.
+- [CVE-2017-1000430](https://github.com/alicemaz/rust-base64/commit/24ead980daf11ba563e4fb2516187a56a71ad319):
+  Integer overflow leading to heap buffer overflow in the `base64` create.
+- [CrosVM](https://bugs.chromium.org/p/chromium/issues/detail?id=892904): Integer
+  overflow leading to heap buffer overflow in ChromeOS's hypervisor.
+- [ring](https://github.com/briansmith/ring/issues/742): Near-miss integer
+  overflow, which would have led to heap buffer overflow.
+
+A panic-by-default behavior would have ensured that none of these bugs would
+have been a vulnerability. All four of these cases involved operations on
+`usize` values. Because Rust is very strict in how integer types are handled, it
+is exceedingly likely that overflows which lead to memory corruption will happen
+on `usize` values: `usize` is consistently used in `core`/`std` APIs which deal
+in lengths, buffer sizes, etc., the kind of values where overflow can be
+dangerous, so it's no coincidence that historic integer-overflow vulnerabilities
+occured with `usize` values.
```

Comment:
As a small thing, I think this should also happen for `isize` -- code that's using `offset` might be calculating in that type instead, and just generally I'd like the two types to be as similar as possible.